### PR TITLE
gtk: add TextBuffer/TextView subclassing support

### PR DESCRIPTION
**Pull request by bilelmoussaoui**
Part of #56

### CODE REVIEW

**Comment by sdroege** [comment]
Let me know when it's ready for review or you have questions :)

File: gtk4/src/subclass/text_view.rs [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+                    granularity.to_glib(),
+                    location.to_glib_none().0,
+                    start.to_glib_none_mut().0,
+                    end.to_glib_none_mut().0,
+                )))
+            } else {
+                glib::signal::Inhibit(false)
+            }
+        }
+    }
```
Comment (bilelmoussaoui) [inline gtk4/src/subclass/text_view.rs:175]: I think this should return glib::Inhibit but not very sure if I've understood that correctly from the C implementation. The corresponding C code is at https://gitlab.gnome.org/GNOME/gtk/-/blob/master/gtk/gtktextview.c#L7070

**Comment by bilelmoussaoui** [comment]
As those two types have a bunch of vfuncs, I will keep the PR this small & work on others separately. Other than the mentioned issue I think the rest should be ok

File: gtk4/src/subclass/text_buffer.rs [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+    end_ptr: *mut ffi::GtkTextIter,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<TextBuffer> = from_glib_borrow(ptr);
+
+    imp.delete_range(
+        &wrap.unsafe_cast_ref(),
+        &mut from_glib_none(start_ptr),
+        &mut from_glib_none(end_ptr),
```
Comment (sdroege) [inline gtk4/src/subclass/text_buffer.rs:403]: This is not sufficient. You need to write back to the pointers afterwards if the iters have changed (also for other mutable text iters in the other functions).

In gtk3 I fixed that somewhere at some point, it's all a bit awful to get right. `gtk_text_iter_assign()` can be useful here.

Comment (sdroege) [reply gtk4/src/subclass/text_buffer.rs:403]: The problem basically is that `from_glib_none()` crates a copy, `from_glib_borrow()` is immutable and not a copy.

File: gtk4/src/subclass/text_view.rs (hunk 1 above)
Comment (sdroege) [reply gtk4/src/subclass/text_view.rs:175]: In GTK4 this is fortunately a bit clearer:

> GDK_EVENT_STOP to stop other handlers from being invoked for the event. GDK_EVENT_PROPAGATE to propagate the event further.

That's exactly `Inhibit`. You can map to `glib::Inhibit` here or alternatively also create a new enum in gdk4 for that.

And you'll want to also properly map such bools to `Inhibit` for signal handlers in `src/auto` etc :)

Comment (bilelmoussaoui) [reply gtk4/src/subclass/text_view.rs:175]: I guess something like https://github.com/gtk-rs/gtk-rs/issues/91 would make sense here as well

File: gtk4/src/subclass/text_buffer.rs (hunk 2 above)
Comment (bilelmoussaoui) [reply gtk4/src/subclass/text_buffer.rs]: Aha right, thanks, should be properly handled now

File: gtk4/src/subclass/text_view.rs (hunk 1 above)
Comment (bilelmoussaoui) [reply gtk4/src/subclass/text_view.rs:175]: I will keep using glib::Inhibit for now here and opened https://github.com/gtk-rs/gtk4-rs/issues/88 to track and fix that issue properly

Comment (sdroege) [reply gtk4/src/subclass/text_view.rs:175]: Makes sense, yes

**Review by sdroege — APPROVED** [review]

**Pull request merged**